### PR TITLE
[JENKINS-30914] Add extra logging to the queue

### PR DESCRIPTION
**Pull request by christ66**
In order to diagnose why something might be blocked in the queue I added loggers in the `Queue.maintain` to understand why a potential item is blocked in the queue and never executes.

@reviewbybees

### CODE REVIEW

File: core/src/main/java/hudson/model/Queue.java [hunk 1]
```diff
@@ -1482,15 +1495,19 @@ public void maintain() {
...
                         updateSnapshot();
                     }
                 } else {
 
                     List<JobOffer> candidates = new ArrayList<JobOffer>(parked.size());
-                    for (JobOffer j : parked.values())
-                        if (j.canTake(p))
+                    for (JobOffer j : parked.values()) {
+                        if (j.canTake(p)) {
+                            LOGGER.log(Level.FINEST, "Potential candidate for ");
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/Queue.java]: for what?

File: core/src/main/java/hudson/model/Queue.java [hunk 2]
```diff
@@ -1595,6 +1616,8 @@ private Runnable makeFlyWeightTaskBuildable(final BuildableItem p){
                 if (n.canTake(p) != null) {
                     continue;
                 }
+
+                LOGGER.log(Level.FINE, "Creating flyweight task {0} for machine {1}", new Object[]{p.task, c.getName()});
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/Queue.java]: "machine". It's a new term... We already have "computer", "node", "slave", etc.

Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/Queue.java]: Why not p.task.getFullDisplayName()?

File: core/src/main/java/hudson/model/Queue.java [hunk 3]
```diff
@@ -1505,10 +1522,12 @@ public void maintain() {
 
                     // found a matching executor. use it.
                     WorkUnitContext wuc = new WorkUnitContext(p);
+                    LOGGER.log(Level.FINER, "Found a matching executor for {0}. Using it.", p.task.getFullDisplayName());
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/Queue.java]: Such approach causes the calculation of  p.task.getFullDisplayName() even if the message won't be printed due to the log level spec

**Comment by oleg-nenashev** [comment]
:bug: Huge number of non-required <code>top.task.getFullDisplayName()</code> calculations during the execution. This method may be implemented by an extension in a plugin, hence the performance impact is unknown.

Just assign a local variable at the beginning of the method.

**Comment by oleg-nenashev** [comment]
:bee:

File: core/src/main/java/hudson/model/Queue.java [hunk 4]
```diff
@@ -1419,7 +1423,13 @@ public void maintain() {
                     Collections.sort(blockedItems, QueueSorter.DEFAULT_BLOCKED_ITEM_COMPARATOR);
                 }
                 for (BlockedItem p : blockedItems) {
+                    String taskDisplayName = p.task.getFullDisplayName();
+                    LOGGER.log(Level.FINER, "Current blocked item: {0}", taskDisplayName);
                     if (!isBuildBlocked(p) && allowNewBuildableTask(p.task)) {
+                        LOGGER.log(Level.INFO,
```
Comment (jtnord) [inline core/src/main/java/hudson/model/Queue.java]: Why is this at info - esp when entering the blocked state is `FINEST`? 
The log levels should be symetrical.

File: core/src/main/java/hudson/model/Queue.java [hunk 5]
```diff
@@ -1438,17 +1448,22 @@ public void maintain() {
...
                 Task p = top.task;
                 if (!isBuildBlocked(top) && allowNewBuildableTask(p)) {
                     // ready to be executed immediately
                     Runnable r = makeBuildable(new BuildableItem(top));
+                    String topTaskDisplayName = top.task.getFullDisplayName();
                     if (r != null) {
+                        LOGGER.log(Level.FINE, "Executing runnable {0}", topTaskDisplayName);
                         r.run();
                     } else {
+                        LOGGER.log(Level.FINEST, "Item {0} was unable to be made a buildable and is now a blocked item.", topTaskDisplayName);
```
Comment (jtnord) [inline core/src/main/java/hudson/model/Queue.java]: and yet above the leaving blocked state is at `INFO` level.

**Comment by jtnord** [comment]
:bug: I believe that the non-symmetrical logging levels is going to cause confusion.

**Comment by christ66** [comment]
@jtnord I've switched to using `Level.FINEST` to log all the queue information.

**Comment by christ66** [comment]
Created JENKINS-30914.

**Comment by jtnord** [comment]
:bee:

**Comment by stephenc** [comment]
:bee:

**Comment by KostyaSha** [comment]
:+1: 
But why not migrate to sl4j? 
About diagnostic, isn't it possible to get queue items and their cause of blockage/internal state? Groovy scripting should allow calling private methods and fields.

**Comment by aheritier** [comment]
@KostyaSha SLF4J and its backward incompatibility between < 1.6 and >= 1.6 is a nightmare to manage. We integrated it in Maven core and I think it was a bad decision. SLF4J is find if you control what you have in your class loader but if it's not the case like in Jenkins or Maven where you have many plugins with various dependencies it is quickly impossible to manage...

**Comment by jtnord** [comment]
@aheritier the [api is backwards compatible](http://www.slf4j.org/faq.html#compatibility) - so I don't understand.  Plugins libs etc should only depend on the API (provided scope) and so long as the implementation is >= this version then it works.  We already have done this in plugins and Jenkins bundles the SLF4J implementation and API and it works (save those broken web containers that leak their version of slf4j onto a webapps classpath).  Maven presumably did something incorrect along the line or have some freaky (incorrect) class loading :-)

**Comment by aheritier** [comment]
@jtnord 

> Plugins libs etc should only depend on the API (provided scope) and so long as the implementation is >= this version then it works

Yes, it is with a IF. because I saw many cases in maven plugins where plugins had a direct or indirect dependency to an old slf4j des and then you are doomed. We are agree, they should only rely on API as provided and have no dependency to any logging impl... But in reality you quickly have a slf4j dep < 1.6 in your claspath

**Comment by KostyaSha** [comment]
@aheritier you may screw core with ANY conflicting version of library from plugin i.e. guava. Then doesn't mean that you shouldn't use guava in core.  If you see bad plugins and scopes, then fill issues to it.

**Comment by christ66** [comment]
> But why not migrate to sl4j?

This might be best in the mailing list more than this PR. I unfortunately do not have much experience with sl4j, so I am unable to provide input on this myself.

> About diagnostic, isn't it possible to get queue items and their cause of blockage/internal state? Groovy scripting should allow calling private methods and fields.

Sometimes I get the cause of blockage as `null` or `???` which does not help diagnosing why it is in the queue. I'm hoping this extra logging information can lead to a timeline of why it does not change the current blocked state.

**Comment by KostyaSha** [comment]
> Sometimes I get the cause of blockage as null

This is valid value http://javadoc.jenkins-ci.org/hudson/model/queue/QueueTaskDispatcher.html#canRun(hudson.model.Queue.Item)

File: core/src/main/java/hudson/model/Queue.java [hunk 6]
```diff
@@ -1385,9 +1385,13 @@ public void maintain() {
                             // we need to clear lost pendings as we cannot know what work unit was on this executor
                             // while it is interrupted. (All this dancing is a result of Executor extending Thread)
                             lostPendings.clear(); // we'll get them next time around when the flag is cleared.
+                            LOGGER.log(Level.FINEST,
+                                    "Interrupt thread for executor {0} is set and we do not know what work unit was on the executor.",
+                                    e.getDisplayName());
```
Comment (jglick) [inline core/src/main/java/hudson/model/Queue.java]: :ant: Just use `getNumber`. Currently you are printing, e.g.,

    Interrupt thread for executor Executor #1 is set and we do not know what work unit was on the executor.

File: core/src/main/java/hudson/model/Queue.java [hunk 7]
```diff
@@ -1419,7 +1423,13 @@ public void maintain() {
                     Collections.sort(blockedItems, QueueSorter.DEFAULT_BLOCKED_ITEM_COMPARATOR);
                 }
                 for (BlockedItem p : blockedItems) {
+                    String taskDisplayName = p.task.getFullDisplayName();
+                    LOGGER.log(Level.FINEST, "Current blocked item: {0}", taskDisplayName);
```
Comment (jglick) [inline core/src/main/java/hudson/model/Queue.java]: Would recommend guarding the potentially nontrivial calculation in

```java
if (LOGGER.isLoggable(Level.FINEST)) {
```

File: core/src/main/java/hudson/model/Queue.java [hunk 8]
```diff
@@ -1438,17 +1448,22 @@ public void maintain() {
...
+                    LOGGER.log(Level.FINEST, "Finished moving all ready items from queue.");
                     break; // finished moving all ready items from queue
+                }
 
                 top.leave(this);
                 Task p = top.task;
                 if (!isBuildBlocked(top) && allowNewBuildableTask(p)) {
                     // ready to be executed immediately
                     Runnable r = makeBuildable(new BuildableItem(top));
+                    String topTaskDisplayName = top.task.getFullDisplayName();
```
Comment (jglick) [inline core/src/main/java/hudson/model/Queue.java]: I..e, `p.getFullDisplayName()` right?

**Comment by jglick** [comment]
Logging framework is off topic here. Jenkins uses `java.util.logging` until otherwise announced.

Some nits but :bee: overall.

File: core/src/main/java/hudson/model/Queue.java [hunk 9]
```diff
@@ -1478,19 +1493,27 @@ public void maintain() {
...
                 } else {
 
                     List<JobOffer> candidates = new ArrayList<JobOffer>(parked.size());
-                    for (JobOffer j : parked.values())
-                        if (j.canTake(p))
+                    for (JobOffer j : parked.values()) {
+                        if (j.canTake(p)) {
+                            LOGGER.log(Level.FINEST,
+                                    "{0} is a potential candidate for task {1}",
+                                    new Object[]{j.executor.getDisplayName(), taskDisplayName});
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/Queue.java]: I'm still not happy about the displayName calculation without Logging level check clause

File: core/src/main/java/hudson/model/Queue.java [hunk 10]
```diff
@@ -1595,6 +1623,8 @@ private Runnable makeFlyWeightTaskBuildable(final BuildableItem p){
                 if (n.canTake(p) != null) {
                     continue;
                 }
+
+                LOGGER.log(Level.FINEST, "Creating flyweight task {0} for computer {1}", new Object[]{p.task.getFullDisplayName(), c.getName()});
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/Queue.java]: doh

**Comment by daniel-beck** [comment]
@oleg-nenashev Three bees and a thumbs up, why do you think this needs review?

**Comment by KostyaSha** [comment]
@daniel-beck i saw that @christ66 extended support-plugin for his case. @christ66 do you still need this loggers?

**Comment by jglick** [comment]
Logging in `support-core` can capture a snapshot of the queue’s contents at the time the bundle is created. This patch should allow people to diagnose why something goes wrong even if the ultimate state does not show it. I think they are complementary.

**Comment by oleg-nenashev** [comment]
I think that @reviewbybees done
Ready for merge if nobody thumbs down due to the minor issues mentioned above

**Comment by abayer** [comment]
Given that no one's spoken up, I'll merge this.

File: core/src/main/java/hudson/model/Queue.java (hunk 7 above)
Comment (jglick) [inline core/src/main/java/hudson/model/Queue.java]: cf. #2791

**Pull request merged**